### ISSUE TITLE
git was not found, check that it is installed and in your PATH env.

### ISSUE THREAD

**Issue by cozylife** [composer/composer#7561]
Output of `composer diagnose`:

```
$ composer diagnose
Checking platform settings: OK
Checking git settings: OK
Checking http connectivity to packagist: OK
Checking https connectivity to packagist: OK
Checking github.com rate limit: OK
Checking disk free space: OK
Checking pubkeys: 
Tags Public Key Fingerprint: 57815BA2 7E54DC31 7ECC7CC5 573090D0  87719BA6 8F3BB723 4E5D42D0 84A14642
Dev Public Key Fingerprint: 4AC45767 E5EC2265 2F0C1167 CBBB8A2B  0C708369 153E328C AD90147D AFE50952
OK
Checking composer version: OK
Composer version: 1.7.2
PHP version: 7.2.8
PHP binary path: /usr/bin/php7.2
```

When I run this command:

```
composer update --prefer-dist --optimize-autoloader --no-dev -vvv
```

I get the following output:

```
Reading ./composer.json
Loading config file ./composer.json
Checked CA file /etc/ssl/certs/ca-certificates.crt: valid
Executing command (/home/forge/site): git branch --no-color --no-abbrev -v
Executing command (/home/forge/site): git describe --exact-match --tags
Executing command (/home/forge/site): git log --pretty="%H" -n1 HEAD
Executing command (/home/forge/site): hg branch
Executing command (/home/forge/site): fossil branch list
Executing command (/home/forge/site): fossil tag list
Executing command (/home/forge/site): svn info --xml
Failed to initialize global composer: Composer could not find the config file: /home/forge/.config/composer/composer.json
To initialize a project, please create a composer.json file as described in the https://getcomposer.org/ "Getting Started" section
Reading /home/forge/site/vendor/composer/installed.json
Running 1.7.2 (2018-08-16 16:57:12) with PHP 7.2.8-1+ubuntu18.04.1+deb.sury.org+1 on Linux / 4.15.0-32-generic
Loading composer repositories with package information
Downloading https://spark-satis.laravel.com/packages.json
Writing /home/forge/.cache/composer/repo/https---spark-satis.laravel.com/packages.json into cache
Reading /home/forge/.cache/composer/repo/https---spark-satis.laravel.com/include-all$5f21269808a937e04c9c3d8574b65688bb3ebf8b.json from cache
Executing command (CWD): git clone --mirror 'git@bitbucket.org:team/project.git' '/home/forge/.cache/composer/vcs/git-bitbucket.org-team-project.git/'
Executing command (CWD): git --version

                                                                                                                                                 
  [RuntimeException]                                                                                                                             
  Failed to clone git@bitbucket.org:team/project.git, git was not found, check that it is installed and in your PATH env.  
                                                                                                                                                 

Exception trace:
 () at phar:///usr/local/bin/composer/src/Composer/Util/Git.php:324
 Composer\Util\Git->throwException() at phar:///usr/local/bin/composer/src/Composer/Util/Git.php:199
 Composer\Util\Git->runCommand() at phar:///usr/local/bin/composer/src/Composer/Util/Git.php:226
 Composer\Util\Git->syncMirror() at phar:///usr/local/bin/composer/src/Composer/Repository/Vcs/GitDriver.php:60
 Composer\Repository\Vcs\GitDriver->initialize() at phar:///usr/local/bin/composer/src/Composer/Repository/VcsRepository.php:88
 Composer\Repository\VcsRepository->getDriver() at phar:///usr/local/bin/composer/src/Composer/Repository/VcsRepository.php:123
 Composer\Repository\VcsRepository->initialize() at phar:///usr/local/bin/composer/src/Composer/Repository/ArrayRepository.php:185
 Composer\Repository\ArrayRepository->getPackages() at phar:///usr/local/bin/composer/src/Composer/DependencyResolver/Pool.php:104
 Composer\DependencyResolver\Pool->addRepository() at phar:///usr/local/bin/composer/src/Composer/Installer.php:378
 Composer\Installer->doInstall() at phar:///usr/local/bin/composer/src/Composer/Installer.php:228
 Composer\Installer->run() at phar:///usr/local/bin/composer/src/Composer/Command/UpdateCommand.php:162
 Composer\Command\UpdateCommand->execute() at phar:///usr/local/bin/composer/vendor/symfony/console/Command/Command.php:241
 Symfony\Component\Console\Command\Command->run() at phar:///usr/local/bin/composer/vendor/symfony/console/Application.php:843
 Symfony\Component\Console\Application->doRunCommand() at phar:///usr/local/bin/composer/vendor/symfony/console/Application.php:193
 Symfony\Component\Console\Application->doRun() at phar:///usr/local/bin/composer/src/Composer/Console/Application.php:254
 Composer\Console\Application->doRun() at phar:///usr/local/bin/composer/vendor/symfony/console/Application.php:117
 Symfony\Component\Console\Application->run() at phar:///usr/local/bin/composer/src/Composer/Console/Application.php:103
 Composer\Console\Application->run() at phar:///usr/local/bin/composer/bin/composer:56
 require() at /usr/local/bin/composer:24
```

And I expected this to happen:
In bash, I can see that git is in the path.
`which git` yields: `/usr/bin/git`

I poured through the composer source code like a wizard and saw it's using PHP exec().
`php -r 'var_dump(exec("/usr/bin/git --version", $out, $ret), $out, $ret);'`

Yields:
```
string(0) ""
array(0) {
}
int(127)
```
That means PHP isn't able to give me the output for the command, thus, Composer cannot "find" Git.

So, I tried going through the PHP source code (code that builds the PHP interpreter) and I saw that it's looking for an input stream being piped in when it runs the exec() and other related functions. So, my guess is it's trying to pipe in the response via stdin instead of running the command directly. I don't know C very well, but I know PHP like the back of my hand.

Anyway, so the issue I need resolved is to find out what's blocking PHP from running the commands on the server when PHP is configured the same way on my local development rig. Here's what I know so far.

1. Stopping apparmor then running the command doesn't make it work.
2. I compared the PHP 7.2 cli ini on the server and local rig and they're similar enough that none of the differences would cause an issue. (They're also from the same Linux repository.
3. Locally I'm on Linux Mint 19, and the server is running Ubuntu 18.04 which is what Linux Mint 19 is based on. They're also running the same Linux Kernel.
4. All software is upgraded including composer.

I know this issue isn't really about Composer in the sense that Composer doesn't have a bug, but some server configuration is causing the issue. (We're on DigitalOcean in case that matters.)

**Comment by johnstevenson** [comment]
> So, I tried going through the PHP source code (code that builds the PHP interpreter) and I saw that it's looking for an input stream being piped in when it runs the exec() and other related functions.

I don't think this is the case. `exec` calls a process with the [r file mode](https://github.com/php/php-src/blob/ac68a423e1676d3191a6fb3870c1cb48765b7fc0/ext/standard/exec.c#L113)

**Comment by cozylife** [comment]
Thanks, @johnstevenson. Any idea what would block PHP from talking to the process?

**Comment by johnstevenson** [comment]
PHP is not being blocked. It is talking to the shell, which returns error code 127 (command not found).

So either `git` is not installed, or it is not in your `PATH` (and the shell cannot find it), which is what the runtime exception is telling you.

**Comment by cozylife** [comment]
@johnstevenson - I appreciate your help on this issue.

Output of `php -i` is:
```
...
$_SERVER['PATH'] => /usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin:/usr/games:/usr/local/games:/snap/bin
...
```
Please note that it includes `/usr/bin`
Output of `which git`
```
/usr/bin/git
```
What am I missing?

**Comment by alcohol** [comment]
Are you running everything as the same (`forge`) user?

**Comment by cozylife** [comment]
Oh, SWEET! Victory is mine! I figured out what was going on and learned something new in the process.

In AppArmor, I used a template as the starting point for the profile and altered it based on the output of `aa-logprof`. However, part of the template I thought nothing about was actually the culprit.

This line blocked access to exec() as apparently PHP uses Dash instead of Bash. (Odd. I've never heard of Dash, so I have to wonder about their reasoning.)
`deny /bin/dash xr, #deny access to dash`

My original hypothesis that stopping AppArmor should allow PHP to do whatever it wants is dead wrong. To fully disable AppArmor for PHP, I needed to restart PHP FPM. (Not sure why that even affected CLI, but it did.)

So, if anyone else comes across this issue, follow these steps.

1. Test the output with `php -r "var_dump(exec('git --version'));"` (If it gives you an output, these steps won't help you. You'll need to find a different solution.)
2. Allow access to /bin/dash for PHP via AppArmor.
3. Do `sudo service apparmor reload`
4. Do `sudo service php7.2-fpm reload` (Might be using a different service name on your server - check it's name with `sudo service --status-all` to find it. Your system might also use something other than `service` and you'll need to locate what one your system uses.)
5. Test the output with `php -r "var_dump(exec('git --version'));"` (If it gives you an output, you're good to go.)

If the last steps didn't help, then try:
1. Test the output with `php -r "var_dump(exec('git --version'));"`
2. Do `sudo service apparmor stop`
3. Do `sudo service php7.2-fpm reload`
4. Test the output with `php -r "var_dump(exec('git --version'));"` (If it gives you an output, you know something in your AppArmor profile is blocking access.)

**Comment by johnstevenson** [comment]
Glad you have found a solution.

> apparently PHP uses Dash instead of Bash.

PHP uses the system `popen`, which invokes whatever shell is defined as the default ([man-pages](http://man7.org/linux/man-pages/man3/popen.3.html))

**Comment by cozylife** [comment]
For anyone who would like to know more about why Ubuntu switched to Dash as the default shell, see: https://lwn.net/Articles/343924/

**TL;DR**
Apparently Dash is faster than Bash. Aside from the login shell, the system defaults to Dash.

**Comment by SamadkhanAshraf101** [comment]
when i run this command,
composer global require laravel/installer
i get the following output

  Failed to download symfony/process from dist: The zip extension and unzip/7z commands are both missing, skipping.
The php.ini used by your command-line PHP is: C:\xampp\php\php.ini
    Now trying to download from source

In GitDownloader.php line 82:

  git was not found in your PATH, skipping source download
